pkg/apis/istio/v1alpha1/istioinstaller_types.proto

```diff
@@ -0,0 +1,461 @@
+    // +optional
+    string clusterIP = 3;
+
+    // externalIPs is a list of IP addresses for which nodes in the cluster
+    // will also accept traffic for this service.  These IPs are not managed by
+    // Kubernetes.  The user is responsible for ensuring that traffic arrives
+    // at a node with this IP.  A common example is external load-balancers
+    // that are not part of the Kubernetes system.
+    // +optional
+    repeated string externalIPs = 5;
+
+    // Supports "ClientIP" and "None". Used to maintain session affinity.
+    // Enable client IP based session affinity.
+    // Must be ClientIP or None.
+    // Defaults to None.
+    // More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies
+    // +optional
+    string sessionAffinity = 7;
+
+    // LoadBalancer will get created with the IP specified in this field.
+    // This feature depends on whether the underlying cloud-provider supports specifying
+    // the loadBalancerIP when a load balancer is created.
+    // This field will be ignored if the cloud-provider does not support the feature.
+    // +optional
+    string loadBalancerIP = 8;
+
+    // If specified and supported by the platform, this will restrict traffic through the cloud-provider
+    // load-balancer will be restricted to the specified client IPs. This field will be ignored if the
+    // cloud-provider does not support the feature."
+    // More info: https://kubernetes.io/docs/tasks/access-application-cluster/configure-cloud-provider-firewall/
+    // +optional
+    repeated string loadBalancerSourceRanges = 9;
+
+    // externalName is the external reference that kubedns or equivalent will
+    // return as a CNAME record for this service. No proxying will be involved.
+    // Must be a valid RFC-1123 hostname (https://tools.ietf.org/html/rfc1123)
+    // and requires Type to be ExternalName.
+    // +optional
+    string externalName = 10;
+
+    // externalTrafficPolicy denotes if this Service desires to route external
+    // traffic to node-local or cluster-wide endpoints. "Local" preserves the
+    // client source IP and avoids a second hop for LoadBalancer and Nodeport
+    // type services, but risks potentially imbalanced traffic spreading.
+    // "Cluster" obscures the client source IP and may cause a second hop to
+    // another node, but should have good overall load-spreading.
+    // +optional
+    string externalTrafficPolicy = 11;
+}
+
+// OperatorConfig defines a config used to install an external operator.
+message OperatorSpec {
+    // manifest_path is the URL for the operator install manifest.
+    string manifest_path = 1;
+
+    // namespace is the namespace the operator manifest and CR are installed into.
+    string namespace = 2;
+
+    // spec is the initial CR for the operator.
+    k8sObjectOverlay spec = 100;
+}
+
+// Patch for an existing k8s resource.
+message k8sObjectOverlay {
```

Comment:
This seems like it could really bloat the CR.

Is there a benefit to using an overlay vs customized templates?  If the idea is to save users from modifying the templates, is there a way the overlays could be added to the templates and applied automatically if they exist (e.g. maybe an overlays folder alongside the templates folder in the chart directory).

I'm still having a hard time understanding the value proposition here.

Reply:
Yes, we want to avoid customizing the templates. I don't follow what you mean by adding overlays to the templates - it seems pretty complex and I think overlays are a user customization so it's easier to apply them to the rendered output than the templates. We as developers will continue to work with the templates, but hopefully less since this mechanism allows users to do customizations without asking us to add parameters to the templates.
Take a look at the overlay example I provided in the header - we're looking for a really simple mechanism to override values that previously required template parameter additions.

Reply:
Here's what I meant regarding overlays and templates.

Take the typical helm structure:
```yaml
base-chart:
  templates: # templates for the main chart
  - resource.yaml # a template for some resource, e.g. deployment maybe
  charts: # directory containing subcharts
    subchart-1: # some subchart
      templates:  # templates for subchart-1
      - resource.yaml # a template for some resource
```

Could we modify that so we have something like:
```yaml
base-chart:
  templates: # templates for the main chart
  - resource.yaml # a template for some resource, e.g. deployment maybe
  overlays: # kustomize overlays for main chart
  - overlay.yaml # some overlay to be applied
  charts: # directory containing subcharts
    subchart-1: # some subchart
      templates:  # templates for subchart-1
      - resource.yaml # a template for some resource
      overlays: # kustomize overlays for subchart-1
      - overlay.yaml # some overlay to be applied
```

This would save users from having to describe pathing, etc. for overrides, etc.  Even if the user just specified a base directory that mirrored the above layout for just the overlays, I think that would be easier to manage.

Another alternative would be to have users layout files in a manner that is identical to kustomize, then just point to that directory.  This would allow you to point users to that documentation, saving them from having to learn the "custom" istio way.

Reply:
We should investigate supporting kustomize because that's that path that the new installer is taking. In all likelihood, we may end up with two ways to overlay - for simple stuff we could embed that in the CRD here to spare the user from creating and editing files, for heavier customization we can support the full blown kustomize framework. It seems to me like a huge burden to place on the user to create files and directory layouts just to override an environment flag (e.g. pod debug level). See https://github.com/istio/operator/pull/8 for some examples.

Reply:
@ostromart, aside from creating a customized operator, users could mount a volume on the operator which contained the overlays, making it rather simple to customize things (e.g. a git volume).

Reply:
Yes, kustomize dir layout could be supported as an alternative to the overlay built into the CRD. I think we should support both approaches and which one is used depends on the use case - for just changing a flag, it would be in the CR, for more extensive overlaying through files and full blown kustomize syntax, it could be the kustomize style structure.